build/result.go

```diff
@@ -22,166 +20,264 @@ import (
+		ch := make(chan *client.SolveStatus)
+		go func() {
+			for {
+				s, ok := <-ch
+				if !ok {
+					return
+				}
+				select {
+				case <-baseCh:
+					// base channel is closed, discard status messages
+				default:
+					baseCh <- s
+				}
 			}
-			select {
-			case statusChan <- s:
-			case <-ctx.Done():
+		}()
+
+		var res *gateway.Result
+		var err error
+		resp, err = cc.Build(ctx, opt, product, func(ctx context.Context, c gateway.Client) (*gateway.Result, error) {
+			var err error
+			res, err = buildFunc(ctx, c)
+
+			if res != nil && err == nil {
+				// Force evaluation of the build result (otherwise, we likely
+				// won't get a solve error)
+				def, err2 := getDefinition(ctx, res)
+				if err2 != nil {
+					return nil, err2
+				}
+				res, err = evalDefinition(ctx, c, def)
 			}
-		}
-	}()
 
-	// get result
-	resultCtxCh := make(chan *ResultContext)
-	errCh := make(chan error)
-	go func() {
-		solveOpt := solveOpt
-		solveOpt.Ref = ""
-		buildDoneCh := make(chan struct{})
-		_, err := c.Build(context.Background(), solveOpt, "buildx", func(ctx context.Context, c gateway.Client) (*gateway.Result, error) {
-			doneErr := errors.Errorf("done")
-			ctx, cancel := context.WithCancelCause(ctx)
-			defer cancel(doneErr)
-
-			// force evaluation of all targets in parallel
-			results := make(map[*pb.Definition]*gateway.Result)
-			resultsMu := sync.Mutex{}
-			eg, egCtx := errgroup.WithContext(ctx)
-			targets.EachRef(func(def *pb.Definition) error {
-				eg.Go(func() error {
-					res2, err := c.Solve(egCtx, gateway.SolveRequest{
-						Evaluate:   true,
-						Definition: def,
-					})
-					if err != nil {
-						return err
-					}
-					resultsMu.Lock()
-					results[def] = res2
-					resultsMu.Unlock()
-					return nil
-				})
-				return nil
-			})
-			resultCtx := ResultContext{}
-			if err := eg.Wait(); err != nil {
+			if err != nil {
+				// Scenario 1: we failed to evaluate a node somewhere in the
+				// build graph.
+				//
+				// In this case, we construct a ResultContext from this
+				// original Build session, and return it alongside the original
+				// build error. We then need to keep the gateway session open
+				// until the caller explicitly closes the ResultContext.
+
 				var se *errdefs.SolveError
 				if errors.As(err, &se) {
-					resultCtx.solveErr = se
-				} else {
-					return nil, err
+					resCtx = &ResultContext{
+						done:     make(chan struct{}),
+						solveErr: se,
+						gwClient: c,
+						gwCtx:    ctx,
+					}
+					respErr = se
+					close(baseCh)
+					close(done)
+					select {
+					case <-resCtx.done:
+					case <-ctx.Done():
+					}
 				}
 			}
-			res2, _ := result.ConvertResult(targets, func(def *pb.Definition) (gateway.Reference, error) {
-				if res, ok := results[def]; ok {
-					return res.Ref, nil
-				}
-				return nil, nil
-			})
+			return res, err
+		}, ch)
+		select {
+		case <-done:
+			cancel(context.Canceled)
+			return
+		default:
+		}
+		close(baseCh)
```

Comment:
How is this not a close of already closed channel.

Reply:
See above, hopefully the new comments and logic make a bit more sense?